libraries/wasm/src/doc.rs

```diff
@@ -0,0 +1,260 @@
+
+        Ok(NewDoc { doc, key })
+    }
+
+    #[wasm_bindgen(constructor)]
+    pub fn new(params: &JsValue) -> Result<Doc, JsValue> {
+        if params.is_object() {
+            let params: DocParams = params.into_serde().map_err(js_err)?;
+
+            let did: IotaDID = if let Some(ref did) = params.did {
+                IotaDID::parse(did).map_err(js_err)?
+            } else {
+                Self::create_did(&params.key)?
+            };
+
+            Self::create(did, params.key, params.tag.as_deref())
+        } else if let Some(public) = params.as_string() {
+            Self::create(Self::create_did(&public)?, public, None)
+        } else {
+            panic!("Invalid Arguments for `new Doc(..)`");
+        }
+    }
+
+    #[wasm_bindgen(getter)]
+    pub fn did(&self) -> String {
+        self.0.did().to_string()
+    }
+
+    #[wasm_bindgen(getter)]
+    pub fn auth_chain(&self) -> String {
+        self.0.did().create_address_hash()
+    }
+
+    #[wasm_bindgen(getter)]
+    pub fn diff_chain(&self) -> String {
+        self.0.diff_address_hash()
+    }
+
+    #[wasm_bindgen(getter)]
+    pub fn authentication_key(&self) -> String {
+        encode_b58(&self.0.authentication_key_bytes())
+    }
+
+    #[wasm_bindgen(getter)]
+    pub fn document(&self) -> JsValue {
+        JsValue::from_serde(&self.0).ok().unwrap_or(JsValue::NULL)
+    }
+
+    #[wasm_bindgen(getter)]
+    pub fn proof(&self) -> JsValue {
+        self.0
+            .metadata()
+            .get("proof")
+            .and_then(|value| JsValue::from_serde(value).ok())
+            .unwrap_or(JsValue::NULL)
+    }
+
+    #[wasm_bindgen]
+    pub fn sign(&mut self, key: &Key) -> Result<JsValue, JsValue> {
+        self.0.sign(key.0.secret()).map_err(js_err).map(|_| JsValue::NULL)
+    }
+
+    #[wasm_bindgen]
+    pub fn verify(&self) -> bool {
+        self.0.verify().is_ok()
+    }
+
+    #[wasm_bindgen]
+    pub fn diff(&self, other: &Doc, key: &Key) -> Result<JsValue, JsValue> {
+        let doc = DIDDocument::from_str(&other.0.to_string()).map_err(js_err)?;
+        let diff: DIDDiff = self.0.diff(doc, key.0.secret()).map_err(js_err)?;
+
+        JsValue::from_serde(&diff).map_err(js_err)
+    }
+
+    #[wasm_bindgen]
+    pub fn verify_diff(&self, diff: String) -> bool {
+        match DIDDiff::from_json(&diff) {
+            Ok(diff) => self.0.verify_diff(&diff).is_ok(),
+            Err(_) => false,
+        }
+    }
+
+    #[wasm_bindgen]
+    pub fn update_service(&mut self, did: WasmDID, url: String, service_type: String) -> Result<Doc, JsValue> {
+        let endpoint = ServiceEndpoint::Url(url.parse().map_err(js_err)?);
+        let service = ServiceBuilder::default()
+            .id(DID::parse(did.0).map_err(js_err)?)
+            .service_type(service_type)
+            .endpoint(endpoint)
+            .build()
+            .map_err(js_err)?;
+        self.0.update_service(service);
+        Ok(Doc(self.0.clone()))
```

Comment:
Any reason this modifies the current document AND returns a *new* document? (same with functions below)

It might be nicer to `clone` first and leave the original document unmodified

Reply:
Right, thanks for the suggestion